File: services/pipeline-rust/src/jobs.rs
use anyhow::{ensure, Context, Result};
use aws_sdk_dynamodb::{
    error::ProvideErrorMetadata,
    types::{AttributeValue as A, TransactWriteItem, Update},
};
use serde_json::{json, Value};
use std::collections::HashMap;

pub fn terminal(status: &str) -> bool {
    matches!(
        status,
        "CACHE_HIT" | "SUCCEEDED" | "FAILED" | "TIMED_OUT" | "ABORTED"
    )
}
pub fn now() -> String {
    chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, true)
}
fn attribute(value: &Value) -> A {
    match value {
        Value::Null => A::Null(true),
        Value::Bool(b) => A::Bool(*b),
        Value::Number(n) => A::N(n.to_string()),
        Value::String(s) => A::S(s.clone()),
        Value::Array(a) => A::L(a.iter().map(attribute).collect()),
        Value::Object(o) => A::M(o.iter().map(|(k, v)| (k.clone(), attribute(v))).collect()),
    }
}
fn value(a: &A) -> Result<Value> {
    Ok(match a {
        A::Null(_) => Value::Null,
        A::Bool(b) => json!(b),
        A::N(n) => serde_json::from_str(n)?,
        A::S(s) => json!(s),
        A::L(a) => Value::Array(a.iter().map(value).collect::<Result<_>>()?),
        A::M(o) => Value::Object(
            o.iter()
                .map(|(k, v)| Ok((k.clone(), value(v)?)))
                .collect::<Result<_>>()?,
        ),
        _ => anyhow::bail!("unsupported job attribute"),
    })
}
fn attrs(value: &Value) -> HashMap<String, A> {
    value
        .as_object()
        .expect("internal object")
        .iter()
        .map(|(k, v)| (k.clone(), attribute(v)))
        .collect()
}
fn key(job: &str) -> HashMap<String, A> {
    attrs(&json!({"PK":format!("JOB#{job}"),"SK":"META"}))
}

pub struct Jobs {
    pub client: aws_sdk_dynamodb::Client,
    pub table: String,
}
impl Jobs {
    pub async fn get(&self, job: &str) -> Result<Option<Value>> {
        let response = self
            .client
            .get_item()
            .table_name(&self.table)
            .set_key(Some(key(job)))
            .consistent_read(true)
            .send()
            .await?;
        response.item.map(|item| value(&A::M(item))).transpose()
    }
    pub async fn update(&self, job: &str, attributes: Value, require_active: bool) -> Result<()> {
        let mut names = HashMap::new();
        let mut values = HashMap::new();
        let mut setters = Vec::new();
        for (i, (name, val)) in attributes
            .as_object()
            .context("invalid update attributes")?
            .iter()
            .enumerate()
        {
            names.insert(format!("#f{i}"), name.clone());
            values.insert(format!(":v{i}"), attribute(val));
            setters.push(format!("#f{i} = :v{i}"));
        }
        names.insert("#updated".into(), "updated_at".into());
        values.insert(":now".into(), A::S(now()));
        setters.push("#updated = :now".into());
        let mut call = self
            .client
            .update_item()
            .table_name(&self.table)
            .set_key(Some(key(job)))
            .update_expression(format!("SET {}", setters.join(", ")))
            .set_expression_attribute_names(Some(names));
        if require_active {
            values.insert(":false".into(), A::Bool(false));
            call=call.condition_expression("attribute_exists(PK) AND (attribute_not_exists(slot_released) OR slot_released = :false)");
        } else {
            call = call.condition_expression("attribute_exists(PK)");
        }
        call.set_expression_attribute_values(Some(values))
            .send()
            .await?;
        Ok(())
    }
    pub async fn release(&self, job: &str, status: &str, attributes: Value) -> Result<bool> {
        ensure!(terminal(status), "cannot release a nonterminal job");
        let Some(current) = self.get(job).await? else {
            return Ok(false);
        };
        if current["slot_released"] == true {
            return Ok(false);
        }
        let user = crate::contract::string(&current, "user_id")?;
        let timestamp = now();
        let mut names = HashMap::from([("#status".into(), "status".into())]);
        let mut values =
            attrs(&json!({":status":status,":true":true,":false":false,":now":timestamp}));
        let mut setters = vec![
            "#status = :status".to_string(),
            "slot_released = :true".into(),
            "updated_at = :now".into(),
        ];
        for (i, (name, val)) in attributes
            .as_object()
            .context("invalid release attributes")?
            .iter()
            .enumerate()
        {
            names.insert(format!("#f{i}"), name.clone());
            values.insert(format!(":v{i}"), attribute(val));
            setters.push(format!("#f{i} = :v{i}"));
        }
        let job_update=Update::builder().table_name(&self.table).set_key(Some(key(job))).update_expression(format!("SET {}",setters.join(", "))).condition_expression("attribute_exists(PK) AND (attribute_not_exists(slot_released) OR slot_released = :false)").set_expression_attribute_names(Some(names)).set_expression_attribute_values(Some(values)).build()?;
        let counter = Update::builder()
            .table_name(&self.table)
            .set_key(Some(attrs(
                &json!({"PK":format!("USER#{user}"),"SK":"COUNTER"}),
            )))
            .update_expression("SET active_count = active_count - :one, updated_at = :now")
            .condition_expression("active_count > :zero")
            .set_expression_attribute_values(Some(attrs(
                &json!({":one":1,":zero":0,":now":timestamp}),
            )))
            .build()?;
        match self
            .client
            .transact_write_items()
            .transact_items(TransactWriteItem::builder().update(job_update).build())
            .transact_items(TransactWriteItem::builder().update(counter).build())
            .send()
            .await
        {
            Ok(_) => Ok(true),
            Err(error)
                if error
                    .as_service_error()
                    .is_some_and(|e| e.code() == Some("TransactionCanceledException")) =>
            {
                if self
                    .get(job)
                    .await?
                    .is_some_and(|v| v["slot_released"] == true)
                {
                    Ok(false)
                } else {
                    Err(error.into())
                }
            }
            Err(error) => Err(error.into()),
        }
    }
    pub async fn scan(&self, maximum: usize) -> Result<Vec<Value>> {
        let mut records = Vec::new();
        let mut next = None;
        while records.len() < maximum {
            let response = self
                .client
                .scan()
                .table_name(&self.table)
                .filter_expression("begins_with(PK, :prefix)")
                .expression_attribute_values(":prefix", A::S("JOB#".into()))
                .limit((maximum - records.len()).min(100) as i32)
                .set_exclusive_start_key(next)
                .send()
                .await?;
            for item in response.items.unwrap_or_default() {
                records.push(value(&A::M(item))?);
            }
            next = response.last_evaluated_key;
            if next.as_ref().is_none_or(|k| k.is_empty()) {
                break;
            }
        }
        Ok(records)
    }
}

File: services/renderer/src/aqw_char_renderer/jobs.py
"""Atomic DynamoDB job admission, terminal transitions, and slot release."""

from __future__ import annotations

from collections.abc import Mapping
from datetime import UTC, datetime, timedelta
from decimal import Decimal
from typing import Any

import boto3
from boto3.dynamodb.types import TypeDeserializer, TypeSerializer
from botocore.exceptions import ClientError

from aqw_char_renderer.contracts import JobRequest, utc_now

_SERIALIZER = TypeSerializer()
_DESERIALIZER = TypeDeserializer()
TERMINAL_STATUSES = frozenset({"CACHE_HIT", "SUCCEEDED", "FAILED", "TIMED_OUT", "ABORTED"})


def discord_notification_pending(record: Mapping[str, Any]) -> bool:
    """Only explicit Discord admissions may publish to the bot result queue."""
    return (
        record.get("request_origin") == "discord"
        and record.get("result_payload") is not None
        and record.get("result_enqueued_at") is None
    )


def _dynamodb_value(value: Any) -> Any:
    if isinstance(value, float):
        return Decimal(str(value))
    if isinstance(value, Mapping):
        return {key: _dynamodb_value(component) for key, component in value.items()}
    if isinstance(value, list):
        return [_dynamodb_value(component) for component in value]
    if isinstance(value, tuple):
        return [_dynamodb_value(component) for component in value]
    return value


def _values(values: Mapping[str, Any]) -> dict[str, Any]:
    return {key: _SERIALIZER.serialize(_dynamodb_value(value)) for key, value in values.items()}


def _item(values: Mapping[str, Any]) -> dict[str, Any]:
    return {key: _SERIALIZER.serialize(_dynamodb_value(value)) for key, value in values.items()}


class UserLimitExceeded(RuntimeError):
    pass


class JobStore:
    def __init__(self, table_name: str, client: Any | None = None) -> None:
        self.table_name = table_name
        self.client = client or boto3.client("dynamodb")

    def acquire(self, request: JobRequest, maximum_active: int) -> None:
        """Admit an operator/CLI job, including smoke tests and full restarts.

        Discord interactions use the bot's own admission transaction. Keep
        origin outside the render request so copying a Discord job's inputs
        does not also copy permission to send notifications.
        """
        now = utc_now()
        expires_at = int((datetime.now(UTC) + timedelta(days=35)).timestamp())
        job = {
            "PK": f"JOB#{request.job_id}",
            "SK": "META",
            "job_id": request.job_id,
            "user_id": request.discord.user_id,
            "guild_id": request.discord.guild_id or "",
            "channel_id": request.discord.channel_id,
            "request_origin": "cli",
            "status": "QUEUED",
            "slot_released": False,
            "created_at": request.created_at,
            "updated_at": now,
            "expires_at": expires_at,
            "request": request.to_dict(),
        }
        try:
            self.client.transact_write_items(
                TransactItems=[
                    {
                        "Put": {
                            "TableName": self.table_name,
                            "Item": _item(job),
                            "ConditionExpression": "attribute_not_exists(PK)",
                        }
                    },
                    {
                        "Update": {
                            "TableName": self.table_name,
                            "Key": _item(
                                {"PK": f"USER#{request.discord.user_id}", "SK": "COUNTER"}
                            ),
                            "UpdateExpression": (
                                "SET active_count = if_not_exists(active_count, :zero) + :one, "
                                "updated_at = :now"
                            ),
                            "ConditionExpression": (
                                "attribute_not_exists(active_count) OR active_count < :limit"
                            ),
                            "ExpressionAttributeValues": _values(
                                {":zero": 0, ":one": 1, ":limit": maximum_active, ":now": now}
                            ),
                        }
                    },
                ]
            )
        except ClientError as error:
            if error.response.get("Error", {}).get("Code") == "TransactionCanceledException":
                raise UserLimitExceeded(
                    f"User already has {maximum_active} active character jobs"
                ) from error
            raise

    def get(self, job_id: str, *, consistent: bool = True) -> dict[str, Any] | None:
        response = self.client.get_item(
            TableName=self.table_name,
            Key=_item({"PK": f"JOB#{job_id}", "SK": "META"}),
            ConsistentRead=consistent,
        )
        raw = response.get("Item")
        if raw is None:
            return None
        return {key: _DESERIALIZER.deserialize(value) for key, value in raw.items()}

    def mark_execution(self, job_id: str, execution_arn: str) -> None:
        self.client.update_item(
            TableName=self.table_name,
            Key=_item({"PK": f"JOB#{job_id}", "SK": "META"}),
            UpdateExpression="SET execution_arn = :arn, #status = :status, updated_at = :now",
            ConditionExpression="attribute_exists(PK)",
            ExpressionAttributeNames={"#status": "status"},
            ExpressionAttributeValues=_values(
                {":arn": execution_arn, ":status": "PREPARING", ":now": utc_now()}
            ),
        )

    def update_status(self, job_id: str, status: str, **attributes: Any) -> None:
        names = {"#status": "status"}
        values: dict[str, Any] = {":status": status, ":now": utc_now()}
        setters = ["#status = :status", "updated_at = :now"]
        for index, (name, value) in enumerate(sorted(attributes.items())):
            name_token = f"#field{index}"
            value_token = f":value{index}"
            names[name_token] = name
            values[value_token] = value
            setters.append(f"{name_token} = {value_token}")
        self.client.update_item(
            TableName=self.table_name,
            Key=_item({"PK": f"JOB#{job_id}", "SK": "META"}),
            UpdateExpression="SET " + ", ".join(setters),
            ExpressionAttributeNames=names,
            ExpressionAttributeValues=_values(values),
        )

    def release(
        self,
        job_id: str,
        status: str,
        *,
        attributes: Mapping[str, Any] | None = None,
    ) -> bool:
        if status not in TERMINAL_STATUSES:
            raise ValueError(f"Cannot release a nonterminal status: {status}")
        current = self.get(job_id)
        if current is None:
            return False
        if current.get("slot_released") is True:
            return False
        user_id = str(current["user_id"])
        names = {"#status": "status"}
        job_values: dict[str, Any] = {
            ":status": status,
            ":true": True,
            ":false": False,
            ":now": utc_now(),
        }
        setters = [
            "#status = :status",
            "slot_released = :true",
            "updated_at = :now",
        ]
        for index, (name, value) in enumerate(sorted((attributes or {}).items())):
            name_token = f"#field{index}"
            value_token = f":value{index}"
            names[name_token] = name
            job_values[value_token] = value
            setters.append(f"{name_token} = {value_token}")
        try:
            self.client.transact_write_items(
                TransactItems=[
                    {
                        "Update": {
                            "TableName": self.table_name,
                            "Key": _item({"PK": f"JOB#{job_id}", "SK": "META"}),
                            "UpdateExpression": "SET " + ", ".join(setters),
                            "ConditionExpression": (
                                "attribute_exists(PK) AND "
                                "(attribute_not_exists(slot_released) OR slot_released = :false)"
                            ),
                            "ExpressionAttributeNames": names,
                            "ExpressionAttributeValues": _values(job_values),
                        }
                    },
                    {
                        "Update": {
                            "TableName": self.table_name,
                            "Key": _item({"PK": f"USER#{user_id}", "SK": "COUNTER"}),
                            "UpdateExpression": "SET active_count = active_count - :one, updated_at = :now",
                            "ConditionExpression": "active_count > :zero",
                            "ExpressionAttributeValues": _values(
                                {":one": 1, ":zero": 0, ":now": job_values[":now"]}
                            ),
                        }
                    },
                ]
            )
        except ClientError as error:
            if error.response.get("Error", {}).get("Code") == "TransactionCanceledException":
                reread = self.get(job_id)
                if reread is not None and reread.get("slot_released") is True:
                    return False
            raise
        return True

    def mark_result_enqueued(self, job_id: str) -> None:
        self.update_status(job_id, str(self.get(job_id)["status"]), result_enqueued_at=utc_now())

    def scan_reconcilable(self, *, maximum: int = 200) -> list[dict[str, Any]]:
        """Return a bounded set of job records that may need repair."""
        records: list[dict[str, Any]] = []
        start_key: dict[str, Any] | None = None
        while len(records) < maximum:
            arguments: dict[str, Any] = {
                "TableName": self.table_name,
                "FilterExpression": "begins_with(PK, :prefix)",
                "ExpressionAttributeValues": _values({":prefix": "JOB#"}),
                "Limit": min(100, maximum - len(records)),
            }
            if start_key is not None:
                arguments["ExclusiveStartKey"] = start_key
            response = self.client.scan(**arguments)
            records.extend(
                {key: _DESERIALIZER.deserialize(value) for key, value in item.items()}
                for item in response.get("Items", [])
            )
            start_key = response.get("LastEvaluatedKey")
            if not start_key:
                break
        return records[:maximum]
